### PR TITLE
Keep ghost particles around

### PR DESCRIPTION
**Pull request by gassmoeller**
A first stab at the problem raised by @egpuckett that many algorithms projecting particle properties onto the mesh want to use particles from neighboring cells, but they might not be available in our code, because they might be owned by a different processor. 
After discussion with @bangerth I tried to keep the ghost particles as separated from the normal particles as possible to avoid the problem of one process changing particle properties that it does not own. Still the access to ghost particles in the interpolator feels somewhat clumsy at the moment (having to ask for the postprocessor once per cell just does not feel right), but other options would involve changing the interface, which would be incompatible, or handing over a merged version of `particles` and `ghost_particles`, which would be slow. I am happy about suggestions for improvements.

### CODE REVIEW

File: source/particle/interpolator/cell_average.cc [hunk 1]
```diff
@@ -57,8 +64,18 @@ namespace aspect
...
         const types::LevelInd cell_index = std::make_pair(found_cell->level(),found_cell->index());
-        const std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator,
-              typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator> particle_range = particles.equal_range(cell_index);
+
+        std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator,
+            typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator> particle_range;
+
+        if (cell->is_locally_owned())
+          {
+            particle_range = particles.equal_range(cell_index);
```
Comment (bangerth) [inline source/particle/interpolator/cell_average.cc]: I think you can write this more elegantly as
```
  std::pair<...> particle_range = (cell->is_locally_owned()
                                                   ?
                                                   particles.equal_range(cell_index)
                                                   :
                                                   ghost_particles.equal_range(cell_index));
```
This way you could also make it `const`.

File: source/particle/world.cc [hunk 2]
```diff
@@ -535,6 +543,67 @@ namespace aspect
...
+      std::vector<std::set<unsigned int> > vertex_to_neighbor_subdomain(this->get_triangulation().n_vertices());
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+      for (; cell != endc; ++cell)
+        {
+          if (cell->is_ghost())
+            for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+              vertex_to_neighbor_subdomain[cell->vertex_index(v)].insert(cell->subdomain_id());
```
Comment (bangerth) [inline source/particle/world.cc]: I suppose this code is the same as in deal.II's DoFHandler ghost exchange of DoF indices? I wonder if we ought to put this functionality somewhere...

File: source/particle/world.cc [hunk 3]
```diff
@@ -535,6 +543,67 @@ namespace aspect
...
+                }
+
+              if (cell_to_neighbor_subdomain.size() > 0)
+                {
+                  const std::pair< const typename std::multimap<types::LevelInd,Particle <dim> >::iterator,
+                        const typename std::multimap<types::LevelInd,Particle <dim> >::iterator>
+                        particle_range_in_cell = particles.equal_range(std::make_pair(cell->level(),cell->index()));
+
+                  for (std::set<unsigned int>::iterator domain=cell_to_neighbor_subdomain.begin();
+                       domain!= cell_to_neighbor_subdomain.end(); ++domain)
```
Comment (bangerth) [inline source/particle/world.cc]: odd spacing around `!=`

File: source/particle/world.cc [hunk 4]
```diff
@@ -535,6 +543,67 @@ namespace aspect
...
+
+                        ghost_particles_by_domain.insert(std::make_pair(*domain,*particle));
+                    }
+                }
+            }
+        }
+
+      std::vector<std::pair<types::LevelInd, Particle<dim> > > received_ghost_particles;
+      send_recv_particles(ghost_particles_by_domain,
+                          received_ghost_particles);
```
Comment (bangerth) [inline source/particle/world.cc]: Oh, how awesome that you could re-use this function!

**Comment by gassmoeller** [comment]
I adressed the comments, this should be ready for merge. 
Actually, I did not check the code of the DoFHandler class, I just looked for a similar function in deal.II that I could use and did not find one so I wrote it here myself. If you think it is useful, I can create a PR to include it in deal.II and eventually remove it from ASPECT again.

**Comment by bangerth** [comment]
Yes, I think figuring out which other processor shares at least one vertex with the locally owned subdomain would be a useful piece of information worth putting into `namespace GridTools`, along with a multi-map from cells for which at least one vertex is adjacent to a ghost cell, to the subdomain id of that ghost cell. You already have that information, and it would be useful to have this in some sort of generic way because it appears in lots of different contexts.

**Pull request merged**